### PR TITLE
Allow macros to call method returning implicit functions

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Extracted from #5297

**Comment by nicolasstucki** [comment]
I first tried to call the method and then call reflectively the apply method of the lambda. This did not work, the reflective call crashed saying that it cannot call a method marcked as "public" :joy:.

**Comment by liufengyun** [comment]
CI is not happy.

**Comment by nicolasstucki** [comment]
Fixed it

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala [hunk 1]
```diff
@@ -347,6 +347,8 @@ object Splicer {
 
     object Call {
       def unapply(arg: Tree): Option[(RefTree, List[Tree])] = arg match {
+        case Select(Call(fn, args), nme.apply) if defn.isImplicitFunctionType(fn.tpe.widenDealias.finalResultType) =>
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/Splicer.scala]: I was missing this guard.

**Comment by nicolasstucki** [comment]
Fixed it

**Pull request merged**